extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/devui/DatabaseInspector.java

```diff
@@ -0,0 +1,174 @@
+            datasources.add(getDatasource(ds));
+        }
+
+        return datasources;
+    }
+
+    private Datasource getDatasource(String datasource) {
+        AgroalDataSource ads = (AgroalDataSource) checkedDataSources.get(datasource);
+        AgroalDataSourceConfiguration configuration = ads.getConfiguration();
+
+        String jdbcUrl = configuration.connectionPoolConfiguration().connectionFactoryConfiguration().jdbcUrl();
+        boolean isDefault = DataSourceUtil.isDefault(datasource);
+
+        return new Datasource(datasource, jdbcUrl, isDefault);
+    }
+
+    public List<Table> getTables(String datasource) {
+
+        List<Table> tableList = new ArrayList<>();
+        try {
+            AgroalDataSource ads = (AgroalDataSource) checkedDataSources.get(datasource);
+            try (Connection connection = ads.getConnection()) {
+                DatabaseMetaData metaData = connection.getMetaData();
+
+                // Get all tables
+                try (ResultSet tables = metaData.getTables(null, null, "%", new String[] { "TABLE" })) {
+                    while (tables.next()) {
+                        String tableName = tables.getString("TABLE_NAME");
+                        String tableSchema = tables.getString("TABLE_SCHEM");
+
+                        // Get the Primary Keys
+                        List<String> primaryKeyList = new ArrayList<>();
+                        try (ResultSet primaryKeys = metaData.getPrimaryKeys(null, tableSchema, tableName)) {
+                            while (primaryKeys.next()) {
+                                String primaryKeyColumn = primaryKeys.getString("COLUMN_NAME");
+                                primaryKeyList.add(primaryKeyColumn);
+                            }
+                        }
+
+                        // Get columns for each table
+                        List<Column> columnList = new ArrayList<>();
+                        try (ResultSet columns = metaData.getColumns(null, tableSchema, tableName, "%")) {
+                            while (columns.next()) {
+                                String columnName = columns.getString("COLUMN_NAME");
+                                String columnType = columns.getString("TYPE_NAME");
+                                int columnSize = columns.getInt("COLUMN_SIZE");
+                                String nullable = columns.getString("IS_NULLABLE");
+                                columnList.add(new Column(columnName, columnType, columnSize, nullable));
+                            }
+                        }
+                        tableList.add(new Table(tableSchema, tableName, primaryKeyList, columnList));
+                    }
+                }
+            }
+
+        } catch (SQLException ex) {
+            throw new RuntimeException(ex);
+        }
+
+        return tableList;
+    }
+
+    public DataSet executeSQL(String datasource, String sql) {
```

Comment:
do i read it right that the client side can request arbitrary SQL execution here and not just limited to browse the known tables? is that not too broad security wise ?

Reply:
Yes you read it right :)

Reply:
Ok - so shouldn't we restrict this somehow ? Do we really need to expose arbitrary execution all the way out to the client ?

Reply:
I'll replay soon with some updates. We will hide this behind config